numpy/core/arrayprint.py

```diff
@@ -1218,19 +1228,27 @@ def format_float_positional(x, precision=None, unique=True,
                               sign=sign, pad_left=pad_left,
                               pad_right=pad_right, min_digits=min_digits)
 
-
 class IntegerFormat:
-    def __init__(self, data):
+    def __init__(self, data, sign=None):
         if data.size > 0:
-            max_str_len = max(len(str(np.max(data))),
-                              len(str(np.min(data))))
+            data_max = len(str(np.max(data)))
+            data_min = len(str(np.min(data)))
+            max_str_len = max(data_max,
+                              data_min)
+            if (sign == '+' 
+                and (np.all(data >= 0)
+                or (data.ndim > 1 and (data_max == data_min)))):
+                max_str_len += 1
+            elif sign == ' ' and np.all(data >= 0):
+                max_str_len += 1
+            elif sign == ' ':
+                sign = '-'
```

Comment:
```suggestion
            data_max = np.max(data)
            data_max_str_len = len(str(data_max))
            if data_max >= 0 and sign in "+ ":
                data_max_str_len += 1

            max_str_len = max(data_max_str_len,
                              str(len(np.min(data))))   
```
When the maximum value is 0 or positive it needs one additional character when printing, that is supposed to be accounted for.  This doesn't matter if there are negative numbers which take more space, but it does matter if there are some negative numbers!

Maybe you can think of a test that covers where this diverges from what you have?

We don't actually support `sign=None`, so might as well make it `sign="-"` in the signature.